### ISSUE TITLE
[no-unsafe-assignment] False positive of destructuring an `any` array with type guard on its element

### ISSUE THREAD

**Issue by yudai-nkt** [typescript-eslint/typescript-eslint#4912]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

Playground for this reproduction is [here](https://typescript-eslint.io/play#ts=4.6.2&sourceType=module&code=MYewdgzgLgBAlhAWgUwE4hgXhgCgIaoDmAXDAK5gDWYIA7mAJSkGHwQwAMWAfDC1pmwcA3AChRoSLABmIDNnxFSFanUakAbiDgATHjADeomPGm4AgqlR4AngDoEl6zcWEGMAGQe2KdK4DaHAC6DO5GJiaS0DD+AEYEQVh8RGImAL6iaWJAA&rules=N4IgAgLgngDgpgZwMYCcCWMIFpEBs0B2EA9AQPZYCuBCAhgGZxa0IJoDmBAtnESAFwg4KFGRQgAvkA&tsConfig=N4XyA).

```JSON
{
  "rules": {
    "@typescript-eslint/no-unsafe-assignment": "error"
  }
}
```

```TS
const isZero = (arg: unknown): arg is 0 => arg === 0;

const foo = (arg: unknown): void => {
  if (Array.isArray(arg) && isZero(arg[0])) {
    const [bar] = arg;
  }
};
```

```json
{
  "compilerOptions": {}
}
```

**Expected Result**

No error is reported.

**Actual Result**

The assignment `const [bar] = arg;` causes ``Unsafe array destructuring of an `any` array value.``

While this line is indeed an array destructuring of `any` array value, I don't think this is unsafe. The type guard `isZero(arg[0])` guarantees that the first element of `arg` is of type `0`, so is `bar` (and the compiler infers as such).

**Additional Info**

N/A.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.22.0` |
| `@typescript-eslint/parser`        | `5.22.0` |
| `TypeScript`                       | `4.6.3` |
| `ESLint`                           | `8.14.0` |
| `node`                             | `16.13.1` |

**Comment by bradzacher** [comment]
> While this line is indeed an array destructuring of `any` array value... 

Full stop here - we operate solely on the types that TS tells us.
If TS tells us the array is of type `any[]`, we cannot know anything more than this.

This is just a limitation of TS's type model.
Ideally your type guard would refine the type of the array to `[0, ...any]`, but it sadly just leaves it as `any[]`.

**Comment by yudai-nkt** [comment]
You mean TS *knows* that `arg[0]` (and therefore `bar`) is of type `0` but *doesn't tell* that information to TS-ES, right?

I might be asking something weird as I'm not familiar with the internal of TypeScript, but just want to make sure I understand correctly.

**Comment by bradzacher** [comment]
https://www.typescriptlang.org/play?#code/MYewdgzgLgBAlhAWgUwE4hgXhgCgIaoDmAXDAK5gDWYIA7mAJSkGHwQwAMWAfDC1pmwcA3AChRoSLABmIDNnxFSFanUakAbiDgATHjADeomPGm4AgqlR4AngDoEl6zcWEGMAGQe2KdK4DaHAC6DO5GJhF8RGKRAPSxMAB6APzGEZLQMP4ARgRBWFGEMRHxkUmpJgC+opViQA

We ask TS for the type of the thing being destructured - the same thing the playground is doing on line 6. And just like what the playground shows - we are told by TS that the type is `any[]`. This is the information we act on.

There's some separate mechanism TS uses to track the control-flow typing of the first element of the array. I don't know what it is, but I don't think it's something we can easily use for the general case.

**Comment by yudai-nkt** [comment]
> I don't think it's something we can easily leverage.

Got it, thank you for the clarification.
